pyocd/rtos/threadx.py

```diff
@@ -0,0 +1,514 @@
+        LOG.debug("ThreadX: _tx_thread_current_ptr = 0x%08x", self._current_ptr)
+
+        self._system_state = symbolProvider.get_symbol_value(
+            '_tx_thread_system_state')
+        if self._system_state is None:
+            return False
+        LOG.debug("ThreadX: _tx_thread_system_state = 0x%08x",
+                  self._current_ptr)
+
+        self._target.session.subscribe(
+            self.event_handler, Target.Event.POST_FLASH_PROGRAM)
+        self._target.session.subscribe(
+            self.event_handler, Target.Event.POST_RESET)
+
+        return True
+
+    def invalidate(self):
+        self._threads = {}
+
+    def event_handler(self, notification):
+        # Invalidate threads list if flash is reprogrammed.
+        LOG.debug("ThreadX: invalidating threads list: %s" %
+                  (repr(notification)))
+        self.invalidate()
+
+    def _build_thread_list(self):
+        # Read the number of threads.
+        threadCount = self._target_context.read32(self._created_cnt)
+
+        # Build up list of all the threads
+        allThreads = TargetList(self._target_context, self._created_ptr)
+        newThreads = {}
+        for threadBase in allThreads:
+            try:
+                # Reuse existing thread objects if possible.
+                if threadBase in self._threads:
+                    t = self._threads[threadBase]
+
+                    # Ask the thread object to update its state and priority.
+                    t.update_info()
+                else:
+                    t = ThreadXThread(self._target_context, self, threadBase)
+                LOG.debug("Thread 0x%08x (%s)", threadBase, t.name)
+                newThreads[t.unique_id] = t
+            except exceptions.TransferError:
+                LOG.debug(
+                    "TransferError while examining thread 0x%08x", threadBase)
+
+        # Is the number of threads correct?
+        if len(newThreads) != threadCount:
+            LOG.warning("ThreadX: thread count mismatch, expected: {0}, found: {1}".format(
+                threadCount, len(newThreads)))
+
+        # Create fake handler mode thread.
+        if self._target_context.read_core_register('ipsr') > 0:
+            LOG.debug("ThreadX: creating handler mode thread")
+            t = HandlerModeThread(self._target_context, self)
+            newThreads[t.unique_id] = t
+
+        self._threads = newThreads
+
+    def get_threads(self):
+        if not self.is_enabled:
+            return []
+        self.update_threads()
+        return list(self._threads.values())
+
+    def get_thread(self, threadId):
+        if not self.is_enabled:
+            return None
+        self.update_threads()
+        return self._threads.get(threadId, None)
+
+    @property
+    def is_enabled(self):
+        if self._system_state is None:
+            return False
+        return self._target_context.read32(self._system_state) < ThreadXThreadProvider.TX_INITIALIZE_IN_PROGRESS
```

Comment:
Is a less-than comparison correct? Just odd to see < a pattern like 0xF0F0F0F0. If it's correct, maybe add comment on how the comparison works?

Reply:
I will add a comment, the system state takes TX_INITIALIZE_IN_PROGRESS and TX_INITIALIZE_IN_PROGRESS+1 during initialization, and 0 (for cortex-m ports). In other ports, it contains interrupt nesting count.